From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Moritz Sanft <58110325+msanft@users.noreply.github.com>
Date: Thu, 23 Jan 2025 12:30:19 +0100
Subject: [PATCH] genpolicy: support dynamic annotations

This adds support for handling annotations with dynamic keys to
genpolicy. This is necessary for use-cases like GPU containers, where
in-cluster components (i.e. post policy-generation) instrument
containers with annotations with varying keys, like `cdi.k8s.io/vfioXY`,
where `XY` corresponds to a dynamic ID.
---
 src/tools/genpolicy/genpolicy-settings.json |  8 ++++++--
 src/tools/genpolicy/rules.rego              | 18 +++++++++++++-----
 src/tools/genpolicy/src/policy.rs           |  7 ++++++-
 3 files changed, 25 insertions(+), 8 deletions(-)

diff --git a/src/tools/genpolicy/genpolicy-settings.json b/src/tools/genpolicy/genpolicy-settings.json
index f98ae153733569012bee35c8e6f197f621edb178..84e0738f4e1326babad93402eb750268238ab77e 100644
--- a/src/tools/genpolicy/genpolicy-settings.json
+++ b/src/tools/genpolicy/genpolicy-settings.json
@@ -293,7 +293,10 @@
             "CAP_BPF",
             "CAP_CHECKPOINT_RESTORE"
         ],
-        "image_layer_verification" : "none"
+        "image_layer_verification" : "none",
+        "dynamic_annotations": {
+            "^cdi\\.k8s\\.io\\/vfio[0-9]{2}$": "^nvidia.com/gpu=[0-9]+$"
+        }
     },
     "kata_config": {
         "oci_version": "1.1.0",
@@ -321,7 +324,8 @@
                 "^AZURE_TENANT_ID=[A-Fa-f0-9-]*$",
                 "^AZURE_FEDERATED_TOKEN_FILE=/var/run/secrets/azure/tokens/azure-identity-token$",
                 "^AZURE_AUTHORITY_HOST=https://login\\.microsoftonline\\.com/$",
-                "^TERM=xterm$"
+                "^TERM=xterm$",
+                "^PCI_RESOURCE_NVIDIA_COM.*=[a-fA-F0-9:.-]*$"
             ]
         },
         "UpdateInterfaceRequest": {
diff --git a/src/tools/genpolicy/rules.rego b/src/tools/genpolicy/rules.rego
index ab644735ff5076a5dac28a786faa512adbe1e6ec..c520253f16d217e3bce3e5de32f9cfda5ece6e90 100644
--- a/src/tools/genpolicy/rules.rego
+++ b/src/tools/genpolicy/rules.rego
@@ -230,18 +230,26 @@ allow_anno(p_oci, i_oci) if {
 }
 allow_anno(p_oci, i_oci) if {
     print("allow_anno 2: p Annotations =", p_oci.Annotations)
+    print("allow_anno 2: p Dynamic Annotations =", policy_data.common.dynamic_annotations)
     print("allow_anno 2: i Annotations =", i_oci.Annotations)
 
-    i_keys := object.keys(i_oci.Annotations)
-    print("allow_anno 2: i keys =", i_keys)
-
-    every i_key in i_keys {
-        allow_anno_key(i_key, p_oci)
+    every i_key, i_val in i_oci.Annotations {
+        allow_anno_keyval(i_key, i_val, p_oci)
     }
 
     print("allow_anno 2: true")
 }
 
+allow_anno_keyval(i_key, i_val, p_oci) {
+  allow_anno_key(i_key, p_oci)
+}
+
+allow_anno_keyval(i_key, i_val, p_oci) {
+  some p_key, p_val in policy_data.common.dynamic_annotations
+  regex.match(p_key, i_key)
+  regex.match(p_val, i_val)
+}
+
 allow_anno_key(i_key, p_oci) if {
     print("allow_anno_key 1: i key =", i_key)
 
diff --git a/src/tools/genpolicy/src/policy.rs b/src/tools/genpolicy/src/policy.rs
index 107ea5c4a1522df4c62b361343dad07a4808f9a8..154fd5682767118d4cf97537935251c046a9af65 100644
--- a/src/tools/genpolicy/src/policy.rs
+++ b/src/tools/genpolicy/src/policy.rs
@@ -428,9 +428,14 @@ pub struct CommonData {
 
     /// Default capabilities for a privileged container.
     pub privileged_caps: Vec<String>,
+
+    /// Dynamic annotations contains arbitrary metadata for the container.
+    /// It is different to `KataSpec.Annotations` in that it allows dynamic keys *and*
+    /// values, and that they are checked for *all* keys, whereas `Annotations`
+    /// only allows dynamic values, and only checks them for certain keys at all.
+    pub dynamic_annotations: BTreeMap<String, String>,
 }
 
-/// Configuration from "kubectl config".
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct ClusterConfig {
     /// Pause container image reference.
